GoogleUtilities/SceneDelegateSwizzler/GULSceneDelegateSwizzler.m

```diff
@@ -0,0 +1,465 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "TargetConditionals.h"
+
+#import <GoogleUtilities/GULAppEnvironmentUtil.h>
+#import <GoogleUtilities/GULLogger.h>
+#import <GoogleUtilities/GULMutableDictionary.h>
+#import <GoogleUtilities/GULSceneDelegateSwizzler.h>
+#import "../Common/GULLoggerCodes.h"
+#import "Internal/GULSceneDelegateSwizzler_Private.h"
+
+#import <objc/runtime.h>
+
+#if UISCENE_SUPPORTED
+
+API_AVAILABLE(ios(13.0), tvos(13.0))
+typedef void (*GULOpenURLContextsIMP)(id, SEL, UIScene *, NSSet<UIOpenURLContext *> *);
+
+API_AVAILABLE(ios(13.0), tvos(13.0))
+typedef void (^GULSceneDelegateInterceptorCallback)(id<UISceneDelegate>);
+
+// The strings below are the keys for associated objects.
+static char const *const kGULRealIMPBySelectorKey = "GUL_realIMPBySelector";
+static char const *const kGULRealClassKey = "GUL_realClass";
+
+static GULLoggerService kGULLoggerSwizzler = @"[GoogleUtilities/SceneDelegateSwizzler]";
+
+// Since Firebase SDKs also use this for app delegate proxying, in order to not be a breaking change
+// we disable App Delegate proxying when either of these two flags are set to NO.
+
+/** Plist key that allows Firebase developers to disable Scene Delegate Proxying. */
+static NSString *const kGULFirebaseSceneDelegateProxyEnabledPlistKey =
+    @"FirebaseSceneDelegateProxyEnabled";
```

Comment:
Introducing a new Info.plist key actually should go through the API review. As for me, reusing the app delegate flag is a good option. In thins case we will spare developers from taking care of another flag also we will be able to keep the docs mostly unchanged.

@ryanwilson WDYT?

Reply:
+1, I think re-using the app delegate flag is the right option as well. Although it's now unfortunately named along with the AppDelegate, ultimately the purpose is to remove boilerplate for startup which SceneDelegate is now part of.

If there are valid use cases you can think of where a developer would want to disable AppDelegate swizzling but keep SceneDelegate (or vice-versa) we can re-evaluate.

Reply:
Reused AppDelegate flags